core/node/metadata_calculator/src/helpers.rs

```diff
@@ -550,8 +553,83 @@ impl L1BatchWithLogs {
             MerkleTreeMode::Lightweight => HashSet::new(),
         };
 
+        let load_tree_writes_latency = METRICS.start_load_stage(LoadChangesStage::LoadTreeWrites);
+        let mut tree_writes = storage
+            .blocks_dal()
+            .get_tree_writes(l1_batch_number)
+            .await?;
+        if tree_writes.is_none() && l1_batch_number.0 > 0 {
+            // If `tree_writes` are present for the previous L1 batch, then it is expected them to be eventually present for the current batch as well.
+            // Waiting for tree writes should be faster then constructing them, so we wait with a reasonable timeout.
+            let tree_writes_present_for_previous_batch = storage
+                .blocks_dal()
+                .check_tree_writes_presence(l1_batch_number - 1)
+                .await?;
+            if tree_writes_present_for_previous_batch {
+                tree_writes = Self::wait_for_tree_writes(storage, l1_batch_number).await?;
+            }
+        }
+        load_tree_writes_latency.observe();
+
+        let storage_logs = if let Some(tree_writes) = tree_writes {
+            // If tree writes are present in DB then simply use them.
+            let writes = tree_writes.into_iter().map(|tree_write| {
+                let storage_key =
+                    StorageKey::new(AccountTreeId::new(tree_write.address), tree_write.key);
+                TreeInstruction::write(storage_key, tree_write.leaf_index, tree_write.value)
+            });
+            let reads = protective_reads.into_iter().map(TreeInstruction::Read);
+
+            // `writes` and `reads` are already sorted, we only need to merge them.
+            writes
+                .merge_by(reads, |a, b| a.key() <= b.key())
+                .collect::<Vec<_>>()
+        } else {
+            // Otherwise, load writes' data from other tables.
+            Self::extract_storage_logs_from_db(storage, l1_batch_number, protective_reads).await?
+        };
+
+        load_changes_latency.observe();
+
+        Ok(Some(Self {
+            header,
+            storage_logs,
+            mode,
+        }))
+    }
+
+    #[allow(clippy::needless_pass_by_ref_mut)]
```

Comment:
Just to understand: Is this a false positive by the Rust compiler? Could make sense to add a comment if so.

Reply:
AFAIU it might because of https://github.com/rust-lang/rust-clippy/issues/11179

Sort of off-topic but we should consider bumping rustc version one of these days if there is no strong reason to keep it at `2023-08-20`

Reply:
Yes, exactly. I will add a comment so it will look like `#[allow(clippy::needless_pass_by_ref_mut)] // false positive`, it's the same for 3 more functions in our code.